### ISSUE TITLE
The bot does not follow rate limit when polling Repology

### ISSUE THREAD

**Issue by AMDmi3** [renovatebot/renovate#10133]
I've noticed a lot of 429 errors in Repology access log triggered by renovate bot. Yesterdays counts:

```
 254 Renovate Bot (GitHub App 2740)
 375 https://github.com/renovatebot/renovate
```

Please implement ~1RPS rate limit.

**Comment by rarkins** [comment]
I think we only have `concurrentRequestLimit` possible currently. @zharinov I thought we had discussed rate limiting previously but can't find any issue with a quick search. Do you remember anything?

**Comment by viceice** [comment]
ref #9550

**Comment by HonkingGoose** [comment]
I'm temped to mark this as a duplicate of the earlier issue (#9550), and bump the priority of "Adding 429 retry capability" (#9550) to `priority-2-important. What do you think @rarkins?

**Comment by rarkins** [comment]
I think it's related but not the same. We want to avoid getting into 429 situations in the first place. Limiting to concurrency of 1 would hopefully help but also not be a full solution.

**Comment by HonkingGoose** [comment]
> I think it's related but not the same.

I'll keep both issues as-is then. 😉

**Comment by rarkins** [comment]
BTW I think that the bot *quits* after one 429, instead of retrying. So actually adding in retrying would technically increase the total requests sent to Repology, although it would respect the retry-after so wouldn't be all at once.

**Comment by HonkingGoose** [comment]
How about adding a new "rule" like this to our `config:base` default configuration? I think most Renovate end-users extend from our `config:base` in the default Renovate setup, as that's what the Renovate onboarding PR gives them.

```json
{
  "hostRules": [
    {
      "matchHost": "repology.org",
      "concurrentRequestLimit": 1
    }
  ]
}
```

**Comment by HonkingGoose** [comment]
@AMDmi3 I made a pull request with the proposed workaround, which is now merged into the mainline branch. See pull request #10146.

It will take time for this change to go live on the hosted GitHub.com app, as the Renovate maintainers need to update the GitHub hosted app to a version that includes the workaround. Users who self-host Renovate will also take time to upgrade.

In short, it will take some time before you can see the impact of this change. I'm also not sure of the size of the reduction. It might be a small reduction or a huge reduction, it's hard to tell right now.

**Comment by AMDmi3** [comment]
Thank you! I expect it to be sufficient, as switching to sequential requests would impose big enough natural delay.

**Comment by HonkingGoose** [comment]
I'll label this `priority-3-normal` for now, as we still need to fix the underlying problem, but at least we should not be badgering the Repology server that often anymore. 😄 

---

> BTW I think that the bot _quits_ after one 429, instead of retrying. So actually adding in retrying would technically increase the total requests sent to Repology, although it would respect the retry-after so wouldn't be all at once.

@AMDmi3 What do you think about this idea? Would you be happier if Renovate listened to a `retry-after` header, even if that increases the total amount of hits (just better spaced out now).

**Comment by AMDmi3** [comment]
I don't think I set `retry-after`, and not sure if it can be done meaningfully.
I can imagine it as follows: 
- say I'm ok with 1RPS so I set `retry-after: 1`  
- client does 20 parallel requests
- gets one 200 reply and and 19 429s with `retry-after: 1`
- client waits a second and does 19 parallel requests...

IMO the correct behavior for the client would be to just decrease the rate multiplicatively on 429s and remember that rate to not hit the limit again.
Another approach would be to use some pessimistic rate by default and set higher rate manually for individual hosts which require and allow it. I do that in repology-linkchecker and am not hitting any 429s at all, though it does millions requests to thousands of hosts.

**Comment by HonkingGoose** [comment]
I've copy/pasted your comment into another issue that specifically talks about adding "429 retry capability" into Renovate bot, as your perspective is helpful on that issue as well. 😄

**Comment by AMDmi3** [comment]
For the note, I'm not seeing any noticeable decrease in 429s since this was deployed
![repology_errors-year](https://user-images.githubusercontent.com/474217/134954165-9ac07668-b8a0-42ee-8d00-540c2f4c5f45.png)

Judging by logs though, there are no more than 2 reqs/sec now, so it looks like it doesn't do parallel requests, just doing sequential requests fast enough. I think I'm fine with that for now, so closing this issue. Still, having proper RPS limiting or 429 handling would be very desirable.

**Comment by HonkingGoose** [comment]
> For the note, I'm not seeing any noticeable decrease in 429s since this was deployed

Thank you @AMDmi3 for your feedback on the results of the workaround that we implemented in PR #10146!

I'm sorry to hear that the workaround is not really helping you much, according to your graph. 😞 

> Judging by logs though, there are no more than 2 reqs/sec now, so it looks like it doesn't do parallel requests, just doing sequential requests fast enough.

The idea of the workaround was to limit each client that uses `config:base` to 1 request to repology per Renovate run.

So of course we'll still need issue #9550 to be done to really fix the problem.
